kdbg/dbgdriver.h

```diff
@@ -13,18 +13,54 @@
 #include <queue>
 #include <list>
 
-
 class VarTree;
 class ExprValue;
 class ExprWnd;
 class KDebugger;
 class QStringList;
+enum class FlavorEnum;
+using tripleKey = std::tuple<FlavorEnum, QString, QString>;
 
-/**
- * Initialize some constants, usefull for many places.
+enum class FlavorEnum { Default, ATT, Intel };
+
+/*
+ * Now, flavors can be represented by many ways:
+ * As enum values when passing around between functions
+ * In the QComboboxes they can be represented as we like, without caring
+ * about how the flavor is interpreted for gdb
+ * And of course the actual flavor string for gdb.
+ *
+ * So, define a global static tuple that contains the possible values.
  */
-static const char DefaultX86Flavor[] = "att";
-static const char DefaultGenericFlavor[] = "default";
+static const tripleKey flavorsTuple[3]
+{
+    std::make_tuple( FlavorEnum::Default, "Default", "att"),
+    std::make_tuple( FlavorEnum::ATT, "ATT", "att"),
+    std::make_tuple( FlavorEnum::Intel, "Intel", "intel")
+};
```

Comment:
IMHO, it is a wrong decision to try to represent the disassembly flavors with enum values. It does not provide advantages:

- You have to convert back and forth between the enum representation and the string.
- You have to cast between enum and int.
- You lose the readable value in the config file. And you hard-code the numeric value anyway because it is now baked into the config file.

Where are the benefits? Memory requirements? Not a concern. No string processing? Meh. Type safety? Yes, I agree that is a bonus.

At any rate, I would not use tuples to represent this database, because the accesses with `get<0>()`, `get<1>()`, etc, are not readable. A small data structure with meaningful member variables would be preferable. But please do not go this route in the first place.

Also, I would prefer not to have the translation between the internal strings `att`, `intel` and the combobox entries `ATT`, `Intel` at this low level. A better place would be closer to the user interface (the dialogs).

Also, the entry for default should read `default`, not `att`, right?

Uh, and another point I just noticed: This is the header file. Don't put static variables in header files.

Reply:
The only benefit for my change is that Default can be read as att, which is true for gdb after all. And some other things like `if(!m_flavor)` etc.

Anyway, I can insert plain Default, ATT, Intel in the comboboxes and convert them on the fly to lower. But I found convenient that default meant att with no further processing.

Also, can you elaborate on the isTargetx86 and m_flavor issue? The former works with m_target. Unless I fail to see another problem.

Thanks for the constructive comments.

Reply:
> The only benefit for my change is that Default can be read as att, which is true for gdb after all.

When the user says "I want the default", I understand that as a wish "let Gdb do what it likes". But when you implement "the default is att", then it means that KDbg defines the default, not Gdb. But this is problematic because KDbg wants to be agnostic of the architecture. How should KDbg then define the default flavor for an ARM architecture? It cannot be "att" again.

> I can insert plain Default, ATT, Intel in the comboboxes and convert them on the fly to lower.

`att` and `intel` are the tokens that specify the flavor internally (we happen to use the tokens that Gdb uses and that is a reasonable choice). `ATT` and `Intel` are the corresponding presentation. Could there be a reason to translate the latter two if the UI language is not English? Are there flavors in other assembly languages that do not translate in this straight forward manner, for example `armv8` internally, but `ARM-v8` is the presentation? Lower-case translation is not general enough.

Furthermore, even though the translation from `Intel` to `intel` is simple lower-case translation, how do you know that the internal setting `intel` must activate the combobox entry `Intel`? You need an explicit case distinction or a table lookup. Don't try to be too fancy, ultimately, you can't avoid some complexity. You have already implemented a lookup table, but it is at a too low level: the debugger driver does not have to be concerned with the translation between internal tokens and UI presentation.

BTW, would it be possible to represent the default setting internally with the empty string instead of `default`? I would guess that this would simplify some conditionals.

> Also, can you elaborate on the isTargetx86 and m_flavor issue? The former works with m_target. Unless I fail to see another problem.

My bad, I am sorry, I did not look carefully enough. You are right, there is no problem.

Reply:
These are valid concerns indeed. I suspect that gdb won't appreciate the empty string to represent the default flavor. Maybe if kdbg could read the driver's name (ie gdb) it could decide that att is the way to go.

Another idea is to use a use-default-flavor checkbox to disable the flavor functionality making gdb ultimately use it's default flavor.

About other targets and their flavors. If isTargetx86 () is false, nothing is submitted to gdb either way. Even setting intel for an arm executable, it does nothing. It's completely ignored.

I tried to create a decent lookup table with this commit so it could help us. We could even use these values in a big switch case in submitDisassFlavor. But it's not elegant, I know it. 

To sum it up. Which way should we store flavors? Plain QStrings?

Also, your idea about effectiveFlavor is good. Should I proceed with it and see if it helps?

Reply:
> I suspect that gdb won't appreciate the empty string to represent the default flavor.

Then I would say: If the effective flavor turns out to be the empty string, do not submit anything to Gdb. It will then use the default flavor. Right?

> We could even use these values in a big switch case in submitDisassFlavor.
> ...
> Which way should we store flavors? Plain QStrings?

If you represent the flavor internally as QStrings with values (empty string), `att`, `intel`, then you do not need a big switch in `submitDisassFlavor()`, just a check for the empty string.

You *will* need two (big?) switches each in the two dialogs that translate the internal strings to the presentation strings and back. Note that you can save on direction of translation when you use [setItemData()](https://doc.qt.io/qt-5/qcombobox.html#setItemData) to store the internal string with each item so that you can use [itemData()](https://doc.qt.io/qt-5/qcombobox.html#itemData) to extract the internal text corresponding to the presentation text.

> Also, your idea about effectiveFlavor is good. Should I proceed with it and see if it helps?

Generally, yes, I have a gut feeling that it is worthwhile. But I do not want to burden you with work. It's ultimately your draw.